### PR TITLE
deployer controller: ensure phase direction

### PR DESCRIPTION
**Pull request by kargakis**
Found out in https://github.com/openshift/origin/issues/8403

Fixes https://github.com/openshift/origin/issues/8403 in conjunction with #8418

@smarterclayton @ironcladlou @pweil- PTAL

### CODE REVIEW

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/util/util_test.go [hunk 1]
```diff
@@ -184,3 +184,169 @@ func TestSort(t *testing.T) {
...
+			current:  deployapi.DeploymentStatusFailed,
+			next:     deployapi.DeploymentStatusFailed,
+			expected: false,
+		},
+	}
+
+	for _, test := range tests {
+		got := CanTransitionPhase(test.current, test.next)
+		if got != test.expected {
+			t.Errorf("%s: expected %t, got %t", test.expected, got)
```
Comment (soltysh) [inline pkg/deploy/util/util_test.go]: You're missing test name here.

**Comment by smarterclayton** [comment]
Approved

**Comment by ironcladlou** [comment]
What invalid transition was possible prior to this change?

**Comment by kargakis** [comment]
> What invalid transition was possible prior to this change?

See https://github.com/openshift/origin/issues/8403#issuecomment-207304958

File: pkg/deploy/util/util_test.go (hunk 1 above)
Comment (kargakis) [inline pkg/deploy/util/util_test.go]: fixed

**Comment by kargakis** [comment]
Any other comments here?

**Comment by smarterclayton** [comment]
Unit tests failed

**Comment by kargakis** [comment]
unit test fixed

**Comment by soltysh** [comment]
LGTM

**Comment by kargakis** [comment]
extended test failure seems unrelated
```
• Failure [322.818 seconds]
Kubectl client
/data/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/test/e2e/kubectl.go:1146
  Update Demo
  /data/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/test/e2e/kubectl.go:161
    should scale a replication controller [Conformance] [It]
    /data/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/test/e2e/kubectl.go:150

    Apr 11 06:58:15.658: Timed out after 300 seconds waiting for name=update-demo pods to reach valid state

    /data/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/test/e2e/util.go:1469
------------------------------
```

**Comment by kargakis** [comment]
[merge]

**Comment by smarterclayton** [comment]
Did not appear to fix the problem:

https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/14441/artifact/origin/artifacts/test-end-to-end-docker/logs/container-router-1-deploy-deployment.log

**Pull request merged**